### PR TITLE
Avoid out-of-context interrupted lock exceptions

### PR DESCRIPTION
**Pull request by hubertp**
### Pull Request Description

This change is a re-do of old PR https://github.com/enso-org/enso/pull/8270 that has been abandoned a while ago.

The change attempts to address to case of random `InterruptedExceptions` that can result when a task is interrupted while waiting on a lock. In comparison to the original PR, this change introduces `withXYZLock` methods with a `Callable` parameter. As a result we ensure correct acquisition/release semantics, logging and handling of interruptions.

### Important Notes

https://github.com/enso-org/enso/pull/8270 was too much hassle to rebase against latest develop. 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.

### CODE REVIEW

File: engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/Locking.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  void assertWriteCompilationLock();
+
+  /**
+   * Executes `callable` while holding a compilation write lock
+   *
+   * @param where the class requesting the lock
+   * @param callable code to be executed while holding the lock
+   * @return the result of calling `callable` or null, if no result is expected
+   */
+  <T> T withWriteCompilationLock(Class<?> where, Callable<T> callable);
```
Comment (JaroslavTulach) [inline engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/Locking.java:19]: This is what I wanted once, I think!

File: engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/ReentrantLocking.scala [hunk 2]
```diff
@@ -105,8 +84,27 @@ class ReentrantLocking(logger: TruffleLogger) extends Locking {
       "should already held write compilation lock during the operation"
     )
 
-  /** @inheritdoc */
-  override def acquireReadCompilationLock(): Long = {
+  def withWriteCompilationLock[T](where: Class[_], callable: Callable[T]): T = {
+    var lockTimestamp: Long = 0
```
Comment (JaroslavTulach) [inline engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/ReentrantLocking.scala:88]: I do remember we talk about lock ordering in the past. Here we should check that _read compilation lock_ isn't held. But looks like that is done in `acquireWriteCompilationLock`.

**Review by JaroslavTulach — APPROVED** [review]
`withXyz` is something I have always dreamed about!

File: engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/Locking.java (hunk 1 above)
Comment (hubertp) [reply engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/Locking.java:19]: https://github.com/enso-org/enso/pull/8270#discussion_r1393982296 yes.

**Review by 4e6 — APPROVED** [review]

**Pull request merged**